compiler/rustc_expand/src/expand.rs

```diff
@@ -1118,12 +1153,19 @@ impl<'a, 'b> MutVisitor for InvocationCollector<'a, 'b> {
                 self.check_attributes(&expr.attrs);
                 self.collect_bang(mac, expr.span, AstFragmentKind::Expr).make_expr().into_inner()
             } else {
-                ensure_sufficient_stack(|| noop_visit_expr(&mut expr, self));
+                assign_id!(self, &mut expr.id, || {
+                    ensure_sufficient_stack(|| noop_visit_expr(&mut expr, self));
+                });
                 expr
             }
         });
     }
 
+    // This is needed in order to set `lint_node_id` for `let` statements
+    fn visit_local(&mut self, local: &mut P<Local>) {
+        assign_id!(self, &mut local.id, || noop_visit_local(local, self));
+    }
```

Comment:
Looks like this shouldn't be necessary, tracking the `Stmt`'s ID should be enough (https://github.com/rust-lang/rust/pull/87146#issuecomment-881890210), the buffered lint reporter in `compiler/rustc_lint/src/early.rs` does seem to associate lints with `Stmt` IDs.